<?xml version="1.0"?>
<launch>
  <arg name="bag_prefix"                   default="/mnt/log/d435"/>
  <arg name="depth_width"                  default="640"/>
  <arg name="depth_height"                 default="360"/>
  <arg name="depth_min"                    default="0.105"/>
  <arg name="depth_max"                    default="10"/>
  <arg name="depth_hit_threshold"          default="0.95"/>
  <arg name="depth_unknown_color"          default="0"/> <!-- 0 to 255 -->
  <arg name="grid_map_width"               default="200"/>
  <arg name="grid_map_height"              default="200"/>
  <arg name="grid_map_resolution"          default="0.025"/>
  <arg name="grid_map_layer_height"        default="0.3"/>
  <arg name="grid_map_accumulation_weight" default="0.05"/> <!-- 0.004 for 1/255 -->
  <arg name="frame_rate"                   default="30"/>

  <arg name="d435"                         default="true"/>
  <arg name="record"                       default="false"/>

  <group ns="d435" if="$(arg d435)">
    <node pkg="nodelet" type="nodelet" name="realsense2_camera_manager" args="manager" output="screen"/>
    <node pkg="nodelet" type="nodelet" name="realsense2_camera" args="load realsense2_camera/RealSenseNodeFactory realsense2_camera_manager">
      <param name="enable_pointcloud"        type="bool" value="false"/>
      <param name="enable_sync"              type="bool" value="false"/>
      <param name="align_depth"              type="bool" value="false"/>
      <param name="enable_fisheye"           type="bool" value="false"/>
      <param name="enable_imu"               type="bool" value="false"/>
  
      <param name="enable_depth"             type="bool" value="true"/>
      <param name="enable_infra1"            type="bool" value="false"/>
      <param name="enable_infra2"            type="bool" value="false"/>
      <param name="enable_color"             type="bool" value="true"/>
  
      <param name="depth_width"              type="int"  value="$(arg depth_width)"/>
      <param name="depth_height"             type="int"  value="$(arg depth_height)"/>
      <param name="depth_fps"                type="int"  value="$(arg frame_rate)"/>
      <param name="color_width"              type="int"  value="$(arg depth_width)"/>
      <param name="color_height"             type="int"  value="$(arg depth_height)"/>
      <param name="color_fps"                type="int"  value="$(arg frame_rate)"/>
      <param name="base_frame_id"            type="str"  value="d435_link"/>
      <param name="depth_frame_id"           type="str"  value="d435_depth_link"/>
      <param name="depth_optical_frame_id"   type="str"  value="d435_depth_optical"/>
    </node>
  </group>

  <node pkg="opengl_ros" type="depth_image_projector"
    name="depth_image_projector" output="screen">
    <param name="color_frame_id"            value="d435_color_optical"/>
    <param name="depth_frame_id"            value="d435_depth_optical"/>
    <param name="map_frame_id"              value="d435_depth_link"/>
    <param name="tf_wait_duration"          value="0.1"/>

    <param name="colorWidth"                value="$(arg depth_width)"/>
    <param name="colorHeight"               value="$(arg depth_height)"/>
    <param name="depthWidth"                value="$(arg depth_width)"/>
    <param name="depthHeight"               value="$(arg depth_height)"/>

    <param name="minDepth"                  value="$(arg depth_min)"/>
    <param name="maxDepth"                  value="$(arg depth_max)"/>
    <param name="depthHitThreshold"         value="$(arg depth_hit_threshold)"/>
    <param name="unknownDepthColor"         value="$(arg depth_unknown_color)"/>

    <param name="gridMapWidth"              value="$(arg grid_map_width)"/>
    <param name="gridMapHeight"             value="$(arg grid_map_height)"/>
    <param name="gridMapResolution"         value="$(arg grid_map_resolution)"/>
    <param name="gridMapLayerHeight"        value="$(arg grid_map_layer_height)"/>
    <param name="gridMapAccumulationWeight" value="$(arg grid_map_accumulation_weight)"/>

    <param name="threshold_l"               value="0"/>
    <param name="svm_coef_a"                value="0.266602955271"/>
    <param name="svm_coef_b"                value="0.0444656815907"/>
    <param name="svm_intercept"             value="-44.3271851592"/>

    <param name="vertex_shader"             value="$(find opengl_ros_lib)/shader/vs_depth_image_projection.glsl"/>
    <param name="geometry_shader"           value="$(find opengl_ros_lib)/shader/gs_depth_image_projection.glsl"/>
    <param name="fragment_shader"           value="$(find opengl_ros_lib)/shader/fs_depth_image_projection.glsl"/>
    <param name="vertex_shader_scaling"     value="$(find opengl_ros_lib)/shader/vs_passthrough.glsl"/>
    <param name="fragment_shader_scaling"   value="$(find opengl_ros_lib)/shader/fs_rgb_to_occupancy_grid.glsl"/>

    <remap from="~depth_to_color" to="/d435/extrinsics/depth_to_color"/>
    <remap from="color_in"        to="/d435/color/image_raw"/>
    <remap from="depth_in"        to="/d435/depth/image_rect_raw"/>
    <remap from="image_out"       to="/occupancy_grid_image"/>
  </node>

  <node pkg="image_view" type="image_view" name="viewer_depth">
    <remap from="image" to="/d435/depth/image_rect_raw"/>
  </node>

  <node pkg="image_view" type="image_view" name="viewer_color">
    <remap from="image" to="/d435/color/image_raw"/>
  </node>

  <node pkg="image_view" type="image_view" name="viewer_out">
    <remap from="image" to="/occupancy_grid_image"/>
  </node>

  <node pkg="rosbag" type="record" name="logger" if="$(arg d435)"
    args="-o $(arg bag_prefix) 
          --lz4
          -a
          -x /(d435|occupancy_grid_image|viewer_depth|viewer_color|viewer_out)/(.*)
          /d435/color/camera_info
          /d435/color/image_raw
          /d435/depth/camera_info
          /d435/depth/image_rect_raw
          /d435/extrinsics/depth_to_color
          /occupancy_grid_image
    "/>
</launch>
